### PR TITLE
[MRG +1]: fix incorrect input to 3dQwarp

### PR DESCRIPTION
**Pull request by nadkarni-na**
also better info about the non-linear weight
variable renaming
spelling corrections

### CODE REVIEW

File: sammba/registration/struct.py [hunk 1]
```diff
@@ -34,9 +35,35 @@ def anats_to_common(anat_filenames, write_dir, brain_volume,
         Maximal levels for each nonlinear warping iteration. Passed iteratively
         to sammba.externals.nipype.interfaces.afni.Qwarp
 
-    nonlinear_minimal_patch : int, optional
-        Minimal patch for the final nonlinear warp, passed to
+    nonlinear_minimal_patch : list of int, optional
```
Comment (salma1601) [inline sammba/registration/struct.py]: nonlinear_minimal_patches

File: sammba/registration/struct.py [hunk 2]
```diff
@@ -9,9 +9,10 @@
 def anats_to_common(anat_filenames, write_dir, brain_volume,
                     registration_kind='affine',
                     nonlinear_levels=[1, 2, 3],
-                    nonlinear_minimal_patch=75,
-                    convergence=0.005, caching=False, verbose=False,
-                    unifize_kwargs=None, brain_extraction_unifize_kwargs=None):
+                    nonlinear_minimal_patches=[75],
+                    nonlinear_weight=None,
```
Comment (salma1601) [inline sammba/registration/struct.py]: better to use nonlinear_weight_filename I think

File: sammba/registration/struct.py [hunk 3]
```diff
@@ -426,75 +456,32 @@ def anats_to_common(anat_filenames, write_dir, brain_volume,
     out_tstat_warp_head = tstat(in_file=out_tcat.outputs.out_file,
                                 outputtype='NIFTI_GZ')
 
-    ###########################################################################
```
Comment (salma1601) [inline sammba/registration/struct.py]: I am not following: why is this removed ?

File: sammba/registration/struct.py [hunk 4]
```diff
@@ -426,75 +456,32 @@ def anats_to_common(anat_filenames, write_dir, brain_volume,
...
-                suffixed_file = fname_presuffix(
+                out_file = fname_presuffix(
                     centered_head_file,
-                    suffix='_warped{}'.format(n_iter + 3))
-                if n_iter == len(nonlinear_levels):
-                    out_file = os.path.join(write_dir,
-                                            os.path.basename(suffixed_file))
-                else:
-                    out_file = suffixed_file
+                    suffix='_warped{}'.format(n_iter + 1 + len(nonlinear_levels)))
```
Comment (salma1601) [inline sammba/registration/struct.py]: you mean out_file first iteration will be suffixed _warped3 if you entered 2 non linear levels ? I am not following ...

Comment (nadkarni-na) [reply sammba/registration/struct.py]: this is consistent with what was already there, naming files based on the iteration of the loop rather than the user-chosen -maxlev or -minpatch

File: sammba/registration/struct.py (hunk 2 above)
Comment (nadkarni-na) [reply sammba/registration/struct.py]: on two grounds it shouldn't be: 1) in the future this could also be a function passed on to modify the automatically-generated mask, 2) many/all uses of "-filename(s)" in input variables should gradually be removed as they unnecessarily lengthen the names of variables that are inherently/obviously files, and are inconsistent with the format of the flags eventually used with the underlying AFNI commands

File: sammba/registration/struct.py (hunk 1 above)
Comment (nadkarni-na) [reply sammba/registration/struct.py]: ok will be fixed in next commit

File: sammba/registration/struct.py (hunk 3 above)
Comment (nadkarni-na) [reply sammba/registration/struct.py]: there was no purpose to having code just for the second iteration

File: sammba/registration/struct.py (hunk 1 above)
Comment (nadkarni-na) [reply sammba/registration/struct.py]: fixed

File: sammba/registration/struct.py (hunk 2 above)
Comment (salma1601) [reply sammba/registration/struct.py]: > 2) many/all uses of "-filename(s)" in input variables should gradually be removed as they unnecessarily lengthen the names of variables

...filename can be replaced with just _file
>that are inherently/obviously files,

this is obvious to us. It can be a folder, it can be a nibabel image, it can be everything. In the specific case of weight I already misunderstood this type and changed some code (remember we talked about that) to make it take integers and get this code merged ! So no, it is not obvious at all that a weight is a file

>and are inconsistent with the format of the flags eventually used with the underlying AFNI commands

This is not very important, cause end users are not assumed to guess which AFNI function we are using and what are its parameters. nipype wrapping itself differs for many parameters from the AFNI parameter names and more generally from the wrapped interfaces names. What is important is the overall consistency of the tool: if it's called filename here, it must be called filename everywhere

File: sammba/registration/struct.py (hunk 3 above)
Comment (salma1601) [reply sammba/registration/struct.py]: but the warp_file computation is different no ?

File: sammba/registration/struct.py [hunk 5]
```diff
@@ -34,9 +35,35 @@ def anats_to_common(anat_filenames, write_dir, brain_volume,
...
-    nonlinear_minimal_patch : int, optional
-        Minimal patch for the final nonlinear warp, passed to
+    nonlinear_minimal_patches : list of int, optional
+        Minimal patches for the final nonlinear warps, passed to
         sammba.externals.nipype.interfaces.afni.Qwarp
+        
+    nonlinear_weight : str, optional
+        Path to a mask used to weight non-linear registration.
+        
+        Background:
```
Comment (salma1601) [inline sammba/registration/struct.py]: I think that this paragraph should be added as a note. People usually use ? on ipython to see what are the function parameters. Long descriptions are unpractical.

**Comment by nadkarni-na** [comment]
Apologies for the multiple commits, in the meantime there's still a big problem: unless -maxlev is explicitly stated, it gets set as 2 or the last specified value. Using -minpatch does not stop this behavior. I have no idea what could be causing this.

**Comment by salma1601** [comment]
I am about to finish the verbosity PR so that qwarp can be verbose. May be this will help ?

**Comment by salma1601** [comment]
Is this PR something like WIP or you think you will finish it soon ?

**Comment by salma1601** [comment]
>Using -minpatch does not stop this behavior. I have no idea what could be causing this.

Can you please give me a code snippet so that I try to reproduce the bug with zurich data ? I mean what arguments you exactly pass to the `anats_to_common` function

**Comment by nadkarni-na** [comment]
I don't have the Zurich data, but in all cases I have seen on three other sets of data, if -maxlev is not set explicitly, it ends up being set anyway to a previously set value. Using -minpatch does not override this. Yes verbose output would help make this more explicit. I will PM you the data and an example script.

Otherwise I have finished doing any more commits here.

**Comment by salma1601** [comment]
Great. Then I am going to give this a second round review, and try fix the `maxlev` bug.

File: sammba/registration/struct.py (hunk 2 above)
Comment (salma1601) [inline sammba/registration/struct.py]: From what I understand (please correct me if wrong), the weight is a binary mask (0 or 1), which needs to include brain and some scalp. May be naming it as `nonlinear_mask_filename` is better, because weight is not very informative to non AFNI users.

File: sammba/registration/struct.py [hunk 6]
```diff
@@ -34,9 +35,35 @@ def anats_to_common(anat_filenames, write_dir, brain_volume,
...
         to sammba.externals.nipype.interfaces.afni.Qwarp
 
-    nonlinear_minimal_patch : int, optional
-        Minimal patch for the final nonlinear warp, passed to
+    nonlinear_minimal_patches : list of int, optional
+        Minimal patches for the final nonlinear warps, passed to
         sammba.externals.nipype.interfaces.afni.Qwarp
+        
+    nonlinear_weight : str, optional
+        Path to a mask used to weight non-linear registration.
```
Comment (salma1601) [inline sammba/registration/struct.py]: May be you can just add that this mask should contain brain and some scalp

File: sammba/registration/struct.py (hunk 5 above)
Comment (salma1601) [inline sammba/registration/struct.py]: I think it is better to include this as a note, as here for instance
https://github.com/sammba-mri/sammba-mri/blob/master/sammba/io_conversions/bruker_dicom.py#L106

File: sammba/registration/struct.py [hunk 7]
```diff
@@ -34,9 +35,35 @@ def anats_to_common(anat_filenames, write_dir, brain_volume,
...
+        
+        Background:
+        Without this weight mask, for automatic non-linear registration, a mask 
+        is generated from the brain-extracted affines and dilated to include
```
Comment (salma1601) [inline sammba/registration/struct.py]: since nibabel calls the sform, qform etc as `affines`, may be it is better to be more explicit, for instance 'brain-extracted images after a first affine registration pass'

File: sammba/registration/struct.py [hunk 8]
```diff
@@ -387,130 +418,85 @@ def anats_to_common(anat_filenames, write_dir, brain_volume,
...
-        in_files=warped_files,
-        out_file=os.path.join(write_dir, 'warped_1iter_heads.nii.gz'))
-    out_tstat_warp_head = tstat(in_file=out_tcat.outputs.out_file,
-                                outputtype='NIFTI_GZ')
-
-    ###########################################################################
-    # Then iterative registration from a given level to another is achieved.
-    # Note that any level below a patch size of 25 will not be done (see
-    # 3dQwarp help for further detail).
-    # The input transform is the former warp and needs to be concatenated to
```
Comment (salma1601) [inline sammba/registration/struct.py]: I think you removed this step and no more concatenate the first warps to `IDENT`. Is it because you figured out what was the bug here and solved it ?

File: sammba/registration/struct.py [hunk 9]
```diff
@@ -387,130 +418,85 @@ def anats_to_common(anat_filenames, write_dir, brain_volume,
...
+                    base_file=out_tstat_allineated_head.outputs.out_file,
+                    noneg=True,
+                    iwarp=True,
+                    weight=nonlinear_weight,
+                    iniwarp=warp_file,
+                    inilev=0,
+                    maxlev=levels_minpatches[n_iter],
+                    out_file=out_file)
+                
+            if n_iter != 0 and n_iter < len(nonlinear_levels):
```
Comment (salma1601) [inline sammba/registration/struct.py]: here I think you need to use `elif`, that is
``` Python
if n_iter == 0:
    ....
elif n_iter < len(nonlinear_levels):
    ....
else:
    ....
```
The `elif` treatment will be applied only if `0 < n_iter < len(nonlinear_levels)` and the `else` treatment will be applied only if `n_iter >= len(nonlinear_levels)`

File: sammba/registration/struct.py [hunk 10]
```diff
@@ -387,130 +418,85 @@ def anats_to_common(anat_filenames, write_dir, brain_volume,
...
                     noneg=True,
                     iwarp=True,
-                    weight=out_mask_tool.outputs.out_file,
-                    iniwarp=[warp_file],
-                    inilev=inilev,
-                    minpatch=nonlinear_minimal_patch,
+                    weight=nonlinear_weight,
+                    iniwarp=warp_file,
+                    inilev=levels_minpatches[n_iter - 1] + 1,
+                    maxlev=levels_minpatches[n_iter],
```
Comment (salma1601) [inline sammba/registration/struct.py]: same here, you can use `level_or_minpatch` instead of `levels_minpatches[n_iter]`

File: sammba/registration/struct.py [hunk 11]
```diff
@@ -387,130 +418,85 @@ def anats_to_common(anat_filenames, write_dir, brain_volume,
...
+            if n_iter != 0 and n_iter >= len(nonlinear_levels):
+                out_qwarp = qwarp(
+                    in_file=centered_head_file,
+                    base_file=out_tstat_warp_head.outputs.out_file,
+                    noneg=True,
+                    iwarp=True,
+                    weight=nonlinear_weight,
+                    iniwarp=warp_file,
+                    inilev=nonlinear_levels[-1] + 1, # not ideal
+                    minpatch=levels_minpatches[n_iter],
```
Comment (salma1601) [inline sammba/registration/struct.py]: same here, you can use `level_or_minpatch` instead of `levels_minpatches[n_iter]`

File: sammba/registration/struct.py [hunk 12]
```diff
@@ -387,130 +418,85 @@ def anats_to_common(anat_filenames, write_dir, brain_volume,
...
-                    nmi=True,
                     noneg=True,
                     iwarp=True,
-                    weight=out_mask_tool.outputs.out_file,
-                    iniwarp=[warp_file],
-                    inilev=inilev,
-                    minpatch=nonlinear_minimal_patch,
+                    weight=nonlinear_weight,
+                    iniwarp=warp_file,
+                    inilev=levels_minpatches[n_iter - 1] + 1,
```
Comment (salma1601) [inline sammba/registration/struct.py]: so default behaviour is:
 `levels_minpatches = [1, 2, 3]`
qwarp1: inilev=0, maxlev=1, iniwarp=affine
qwarp2: inilev=2, maxlev=2, iniwarp=warp from qwarp1
qwarp3: inilev=3, maxlev=3, iniwarp=warp from qwarp2
Why don't you do from 1 to 2 and from 2 to 3 ??

File: sammba/registration/struct.py [hunk 13]
```diff
@@ -387,130 +418,85 @@ def anats_to_common(anat_filenames, write_dir, brain_volume,
...
+                                                 centered_head_files):
 
+            if n_iter == 0:
+                out_qwarp = qwarp(
+                    in_file=centered_head_file,
+                    base_file=out_tstat_allineated_head.outputs.out_file,
+                    noneg=True,
+                    iwarp=True,
+                    weight=nonlinear_weight,
+                    iniwarp=warp_file,
```
Comment (salma1601) [inline sammba/registration/struct.py]: in nipype Qwarp interface, `iniwarp` should be a list of existing files. Did you check that when you run the function you din't get a trait type error ?

File: sammba/registration/struct.py [hunk 14]
```diff
@@ -387,130 +418,85 @@ def anats_to_common(anat_filenames, write_dir, brain_volume,
...
                     base_file=out_tstat_warp_head.outputs.out_file,
-                    nmi=True,
                     noneg=True,
                     iwarp=True,
-                    weight=out_mask_tool.outputs.out_file,
-                    iniwarp=[warp_file],
-                    inilev=inilev,
-                    minpatch=nonlinear_minimal_patch,
+                    weight=nonlinear_weight,
+                    iniwarp=warp_file,
```
Comment (salma1601) [inline sammba/registration/struct.py]: same here, to check that this does not produce an error

File: sammba/registration/struct.py [hunk 15]
```diff
@@ -387,130 +418,85 @@ def anats_to_common(anat_filenames, write_dir, brain_volume,
...
-                                        outputtype='NIFTI_GZ')
+                    
+            if n_iter != 0 and n_iter >= len(nonlinear_levels):
+                out_qwarp = qwarp(
+                    in_file=centered_head_file,
+                    base_file=out_tstat_warp_head.outputs.out_file,
+                    noneg=True,
+                    iwarp=True,
+                    weight=nonlinear_weight,
+                    iniwarp=warp_file,
```
Comment (salma1601) [inline sammba/registration/struct.py]: to check type

File: sammba/registration/struct.py (hunk 11 above)
Comment (salma1601) [reply sammba/registration/struct.py]: so if a user enters as `nonlinear_minpatches = [75]` and leaves the `nonlinear_levels` to default, we the two last steps will be
qwarp3: inilev=3, maxlev=3, iniwarp=warp from qwarp2
qwarp4: inilev=4, minpatch=75, inilev=warp from qwarp3

So again, why not to perform from inilev 3 to minpatch 75 ?

**Comment by salma1601** [comment]
I finished my review. There are many changes unclear to me, including the new Qwarp strategy.

**Comment by salma1601** [comment]
I didn't succeed to reproduce your bug. I was using the master branch with verbosity and running nonlinear registration with your choices (`nonlinear_levels` = [0, 2, 4] and 'nonlinear_min_patch`=15). This is what I get as stdout
>3dQwarp -base /home/salma/nilearn_data/pauline_rat/template2/nipype_mem/sammba-externals-nipype-interfaces-afni-preprocess-Qwarp/a9ec94fc4fb9532128d3b158490084dc/warped_1iter_heads_tstat.nii.gz -source /home/salma/nilearn_data/pauline_rat/template2/nipype_mem/sammba-externals-nipype-interfaces-afni-preprocess-Qwarp/a9ec94fc4fb9532128d3b158490084dc/MSME_MIRCen_RatBrain_atlas_bf3_1_0_Unifized_for_brain_extraction_resample.nii.gz -inilev 2 -iniwarp /home/salma/nilearn_data/pauline_rat/template2/nipype_mem/sammba-externals-nipype-interfaces-afni-utils-NwarpCat/a11acdd946dff91624b07101358862a3/iniwarp.nii.gz -noneg -prefix /home/salma/nilearn_data/pauline_rat/template2/nipype_mem/sammba-externals-nipype-interfaces-afni-utils-Resample/a637784ec03aad72200732673f104cdf/MSME_MIRCen_RatBrain_atlas_bf3_1_0_Unifized_for_brain_extraction_resample_warped2.nii.gz -verb -weight /home/salma/nilearn_data/pauline_rat/template2/nipype_mem/sammba-externals-nipype-interfaces-afni-utils-MaskTool/17490136c0d790f98e73b93657a04c17/rigid_body_registered_heads_mask_mask.nii.gz -maxlev 4

File: sammba/registration/struct.py (hunk 2 above)
Comment (nadkarni-na) [reply sammba/registration/struct.py]: It is a weight, it does not have to be binary

File: sammba/registration/struct.py (hunk 12 above)
Comment (nadkarni-na) [reply sammba/registration/struct.py]: If you go from 1 to 2 (as opposed to doing only level 1), then level 2 has been done, so there is no point redoing it again

File: sammba/registration/struct.py (hunk 11 above)
Comment (nadkarni-na) [reply sammba/registration/struct.py]: it is safer to defaulting to the use of levels as much as possible since we have not fully determined the image size / level / patch size relationship

**Comment by nadkarni-na** [comment]
this output does not mention -minpatch being used which gives the impression it is ignored and so would count as a bug even within the output itself

I found the bug by running ps -aux in a shell while the last non-linear loop was running

File: sammba/registration/struct.py (hunk 12 above)
Comment (salma1601) [reply sammba/registration/struct.py]: Ah ok, the maxlev is included. Sorry.

File: sammba/registration/struct.py (hunk 11 above)
Comment (salma1601) [reply sammba/registration/struct.py]: Ok

**Comment by salma1601** [comment]
>In addition the verbose output isn't verbose enough to help the user tune the use of levels and minpatch where needed. Qwarp output usually looks like

Yes I am using my verbosity branch so I have maximal verbosity and I am having things like
>171219-19:41:28,494 interface INFO:
	 stderr 2017-12-19T19:41:28.494397: +        cubic patch 153..167 021..037 092..106 : cost=-2.05431 iter=77 : energy=0.386:0.837 pen=0.0150683 pure=-2.06938 

so for me it is going up to patch 21 and I don't see anything like maxlev 2

**Comment by salma1601** [comment]
I am going to run it tomorrow with your own code branch to see if I can reproduce the bug.

File: sammba/registration/struct.py (hunk 13 above)
Comment (salma1601) [reply sammba/registration/struct.py]: actually I checked and it throws an error
>TraitError: The 'iniwarp' trait of a QwarpInputSpec instance must be a list of items which are an existing file name, but a value of u'/tmp/pauline/template11/MSME_MIRCen_RatBrain_atlas_bf5_0_Unifized_for_brain_extraction_resample_shr_affine_catenated.aff12.1D' <type 'unicode'> was specified.

So this must be corrected

File: sammba/registration/struct.py [hunk 16]
```diff
@@ -387,130 +418,85 @@ def anats_to_common(anat_filenames, write_dir, brain_volume,
...
+                    out_file=out_file)
+                    
+            warped_files.append(out_qwarp.outputs.warped_source)
+            warp_files.append(out_qwarp.outputs.source_warp)
+        
+        out_tcat = tcat(
+            in_files=warped_files,
+            out_file=os.path.join(
+			    write_dir,
+				'warped_{0}iters_hetemplate_filenameads.nii.gz'))
```
Comment (salma1601) [inline sammba/registration/struct.py]: 'warped_{0}iters_hetemplate_filenameads.nii.gz'.format(n_iter)

**Comment by salma1601** [comment]
I can't reproduce the bug with your branch neither. Here is the last qwarp command run
>3dQwarp -base /tmp/pauline/template12/nipype_mem/sammba-externals-nipype-interfaces-afni-preprocess-Qwarp/e80d6fbed33ead11d3de22180ba47625/warped_2iters_hetemplate_filenameads_tstat.nii.gz -source /tmp/pauline/template12/nipype_mem/sammba-externals-nipype-interfaces-afni-preprocess-Qwarp/e80d6fbed33ead11d3de22180ba47625/MSME_MIRCen_RatBrain_atlas_bf3_2_Unifized_for_brain_extraction_resample.nii.gz -inilev 7 -iniwarp /tmp/pauline/template12/nipype_mem/sammba-externals-nipype-interfaces-afni-utils-Resample/316369adc95ce71b79b5b199cf4eab0b/MSME_MIRCen_RatBrain_atlas_bf7_3_Unifized_for_brain_extraction_resample_warped2_WARP.nii.gz -iwarp -minpatch 15 -noneg -prefix /tmp/pauline/template12/nipype_mem/sammba-externals-nipype-interfaces-afni-utils-Resample/316369adc95ce71b79b5b199cf4eab0b/MSME_MIRCen_RatBrain_atlas_bf7_3_Unifized_for_brain_extraction_resample_warped3.nii.gz -weight /tmp/pauline/template10/affbrainsunionmaskdil5.nii.gz

as you can see, the inilev is 7, the minpatch is 15 and there is no maxlev

**Comment by salma1601** [comment]
@nadkarni-na There are many corrections to improve this branch code. Are you planning to do them or can I move to merging #67 ?

File: sammba/registration/struct.py (hunk 2 above)
Comment (nadkarni-na) [reply sammba/registration/struct.py]: Done

File: sammba/registration/struct.py (hunk 6 above)
Comment (nadkarni-na) [reply sammba/registration/struct.py]: Done

File: sammba/registration/struct.py (hunk 5 above)
Comment (nadkarni-na) [reply sammba/registration/struct.py]: Done

Comment (nadkarni-na) [reply sammba/registration/struct.py]: Done

File: sammba/registration/struct.py (hunk 8 above)
Comment (nadkarni-na) [reply sammba/registration/struct.py]: I'm in two minds abut this. I didn't note which files generated this bug. But I have not seen it since so am not entirely sure if it ever really was a real bug. Will readd the IDENT NwarpCat anyway as it's only a line and doesn't do any harm

File: sammba/registration/struct.py (hunk 13 above)
Comment (nadkarni-na) [reply sammba/registration/struct.py]: Corrected. Dunno how I never got an error.

File: sammba/registration/struct.py (hunk 9 above)
Comment (nadkarni-na) [reply sammba/registration/struct.py]: Just noting that using elif causes the interpreter to produce an error as it now sees the out_tstat_warp_head.outputs.out_file which does not exist yet, so will create an empty one before the loop

File: sammba/registration/struct.py [hunk 17]
```diff
@@ -435,69 +441,71 @@ def anats_to_common(anat_filenames, write_dir, brain_volume,
...
-                    iniwarp=warp_file,
+                    weight=nonlinear_weight_file,
+                    iniwarp=[warp_file],
                     inilev=levels_minpatches[n_iter - 1] + 1,
-                    maxlev=levels_minpatches[n_iter],
+                    maxlev=level_or_minpatch,
                     out_file=out_file)
                     
-            if n_iter != 0 and n_iter >= len(nonlinear_levels):
+            elif n_iter >= len(nonlinear_levels):
```
Comment (salma1601) [inline sammba/registration/struct.py]: I think this one should be `else`, because it is executed only when the `if` and `elif` conditions fail, so `n_iter` is systematically greater or equal to `len(nonlinear_levels)`

File: sammba/registration/struct.py [hunk 18]
```diff
@@ -435,69 +441,71 @@ def anats_to_common(anat_filenames, write_dir, brain_volume,
...
-
+    out_tstat_warp_head = out_tstat_allineated_head # prevent later elif failure
+    
     for n_iter, level_or_minpatch in enumerate(levels_minpatches):
         
-        if n_iter == 0:
-            previous_warp_files = affine_transform_files
-        else:
-            previous_warp_files = warp_files
+        if n_iter == 0: previous_warp_files = affine_transform_files
```
Comment (salma1601) [inline sammba/registration/struct.py]: restyling
```Python
if n_iter == 0:
    previous_warp_files = affine_transform_files
```

File: sammba/registration/struct.py [hunk 19]
```diff
@@ -387,130 +424,87 @@ def anats_to_common(anat_filenames, write_dir, brain_volume,
...
-    # Note that any level below a patch size of 25 will not be done (see
-    # 3dQwarp help for further detail).
-    # The input transform is the former warp and needs to be concatenated to
-    # IDENT initially; I forget why, I think it is to avoid some weird bug.
-    if len(nonlinear_levels) > 1:
-        previous_warp_files = warp_files
+    if nonlinear_minimal_patches is None:
+       nonlinear_minimal_patches = []
+    levels_minpatches = nonlinear_levels + nonlinear_minimal_patches
+    out_tstat_warp_head = out_tstat_allineated_head # prevent later elif failure
```
Comment (salma1601) [inline sammba/registration/struct.py]: I think we shouldn't care about the interpreter complaint here. I think the case is too much complicated for it to see that `out_tstat_warp_head` will be created before called

**Comment by salma1601** [comment]
2 minor comments, then it is ready for merge I think. Thanks for your work !

File: sammba/registration/struct.py (hunk 19 above)
Comment (nadkarni-na) [reply sammba/registration/struct.py]: on my system it's more than a complaint: it's a stop error so there's no choice

File: sammba/registration/struct.py (hunk 17 above)
Comment (nadkarni-na) [reply sammba/registration/struct.py]: done

File: sammba/registration/struct.py (hunk 18 above)
Comment (nadkarni-na) [reply sammba/registration/struct.py]: done

File: sammba/registration/struct.py (hunk 19 above)
Comment (salma1601) [reply sammba/registration/struct.py]: Ok. Let me just run the code with all your new changes on the rats data to try figure out why it is a stop error.

Comment (salma1601) [reply sammba/registration/struct.py]: @nadkarni-na I run your updated changes without this line and I have no error ! Could you give me the traceback of your error ?

Comment (nadkarni-na) [reply sammba/registration/struct.py]: I cannot reproduce it anymore, I think it was a consequence of an overused terminal window

File: sammba/registration/struct.py [hunk 20]
```diff
@@ -11,7 +11,8 @@ def anats_to_common(anat_filenames, write_dir, brain_volume,
                     nonlinear_levels=[1, 2, 3],
                     nonlinear_minimal_patches=[75],
                     nonlinear_weight_file=None,
-                    convergence=0.005, caching=False, verbose=True,
+                    convergence=0.005, two_blur=1.1,
```
Comment (salma1601) [inline sammba/registration/struct.py]: `two_blur` is not very informative, what about `blur_radius`

Comment (salma1601) [reply sammba/registration/struct.py]: please also add this parameter in doc

**Comment by salma1601** [comment]
@nadkarni-na I added WIP to the PR title, feel free to remove it when you finish working on it

File: sammba/registration/struct.py (hunk 20 above)
Comment (nadkarni-na) [reply sammba/registration/struct.py]: done

**Comment by nadkarni-na** [comment]
I'm finally getting results exactly the same as (as good as) the shell version, at least on one test dataset, so almost there

**Comment by nadkarni-na** [comment]
@salma1601 I've finished all the tests it's good to merge
